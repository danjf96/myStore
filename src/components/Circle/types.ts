import { TextStyle, ViewStyle } from "react-native";

export interface CIRCLEPROPS {
    style?: ViewStyle,
    styleText?: TextStyle,
    number?: number
}