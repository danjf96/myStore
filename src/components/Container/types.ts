import React from "react";
import { ViewProps, ViewStyle } from "react-native";

export interface PROPSCONTAINER extends ViewProps {
    style?: ViewStyle
}