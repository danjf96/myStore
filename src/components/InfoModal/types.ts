import { ImageRequireSource, ModalProps } from "react-native";

export interface PROPSMODALINFO extends ModalProps {
    image?: ImageRequireSource
    title: string,
    text: string,
    textButton?: string,
    onPress: () => void
    textSecondButton?: string,
    onPressSecondButton?: ()=> void
}
