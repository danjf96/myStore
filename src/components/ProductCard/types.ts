import { PRODUCTSTATE } from "../../store/ducks/products/types";

export interface ProductsCardProps extends PRODUCTSTATE {
    number?: number,
    quantity?: boolean,
    onPressQuantity?: (type: 'increase' | 'decrease') => void,
    testId?: string
}