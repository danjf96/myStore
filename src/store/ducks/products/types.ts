export interface STATESTORE {
    loading: boolean,
    listProducts: PRODUCTSTATE[]
}

//TYPES ACTIONS
export const Types = {
    CLEAN: 'CLEAN',
    CHANGE: 'CHANGE_PRODUCT',
    LOAD: 'LOAD_PRODUCT'
}

export interface PRODUCTSTATE {
    id: string,
    createdAt: string,
    name: string,
    price: number,
    image: string,
    stock: number
}
