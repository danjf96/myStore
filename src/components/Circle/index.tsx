import React from 'react'
import { Text, View } from 'react-native'
import { useSelector } from 'react-redux'
import Styles from './styles'
import { CIRCLEPROPS } from './types'

const Circle = (props: CIRCLEPROPS) => {
    const { shoppingCart: { cart } } = useSelector( ({  shoppingCart  } ) => ({ shoppingCart }) )
    const visible = props.number != undefined ? props.number > 0 : cart.length > 0

    return(
        visible ? 
            <View style={{ ...Styles.container,...props.style }}>
                <Text style={{ ...Styles.text, ...props.styleText }}>{props.number || cart.map( c => c.number).reduce( (a,b) => a+b)}</Text>
            </View>
            : null
        
    )
}

export default Circle