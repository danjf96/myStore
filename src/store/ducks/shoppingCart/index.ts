import api, { errorHandling } from "../../../services/api";
import { STATESTORE, Types } from "./types";


//REDUCERS
const INITIAL_STATE: STATESTORE = {
    loading: false,
    cart: []
}

//STATE CASES
const STATE = (state = INITIAL_STATE, action: any):STATESTORE => {
    switch (action.type) {
        case Types.LOAD:
            return { ...state, loading: action.loading }
        case Types.CHANGE:
            return { ...state, ...action.payload }
        case Types.CHANGE_VALUE:
            return { ...state, [action.key]: action.value }
        case Types.CLEAN:
            return INITIAL_STATE
        default:
            return state;
    }
}

//Actions Creators
export const changeShoppingCart = (payload:{}) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.CHANGE, payload  })
    }
}

export const changeShoppingCartValue = (key:string, value:any) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.CHANGE, key, value  })
    }
}
export default STATE