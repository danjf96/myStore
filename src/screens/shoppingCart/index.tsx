import React, { useEffect, useState } from 'react'
import { FlatList, Text, TouchableOpacity, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { GlobalStyle } from '../../assets/GlobalStyles'
import Container from '../../components/Container'
import InfoModal from '../../components/InfoModal'
import ProductCard from '../../components/ProductCard'
import { changeShoppingCart } from '../../store/ducks/shoppingCart'
import { formatMoney } from '../../utils/mask'
import Styles from './styles'

const ShoppingCart = (props:any) => {

    const dispatch = useDispatch()
    const { shoppingCart: { cart, loading, totalPrice } } = useSelector( ({ shoppingCart  } ) => ({  shoppingCart }) )
    const [ openModal, setOpenModal ] = useState(true)

    const changeQuantity = (type: string, id:string) => {
        let newCart = cart
        switch(type){
            case 'increase':
                newCart = newCart.map( n => n.id == id ? { ...n, number: n.number += 1 } : n )
            break;
            case 'decrease':
                if(newCart.find( n => n.id == id && n.number - 1 == 0))
                    newCart = newCart.filter( n => n.id != id)
                else
                    newCart = newCart.map( n => n.id == id ? { ...n, number: n.number -= 1 } : n )
            break
        }
        const totalPrice = newCart.length > 0 ? newCart.map( n => n.price * n.number).reduce( (a,b) => a+b) : 0
        dispatch( changeShoppingCart({ cart: newCart, totalPrice }))

    }

    return (
        <>
            <Container style={{ height: '100%', paddingBottom: 120}}>
                <Text style={Styles.price}> Total: R$ {formatMoney(totalPrice)}</Text>

                <View style={GlobalStyle.line}></View>
                
                <FlatList 
                    renderItem={ ({ item, index}) => <ProductCard 
                            {...item}
                            testId={'cart'}
                            quantity={true}
                            styleContainer={Styles.productContainer}
                            onPressQuantity={ t => changeQuantity(t, item.id)}
                        />
                    }
                    keyExtractor={ (item,index) => `cart${index}`}
                    data={cart}
                />   

                <InfoModal  
                    testID='finishModal'
                    visible={openModal}
                    title={'Finalizar compra!'} 
                    text={'Escolha uma forma de pagamento'} 
                    textButton={'USAR CARTÃƒO CADASTRADO'} 
                    textSecondButton='USAR SALDO DISPONIVEL' 
                    onPress={() => null} 
                    onPressSecondButton={() => null}
                    onClose={ () => setOpenModal(false)}
                />  

            </Container>

            <View style={Styles.footer}>
                <TouchableOpacity style={Styles.buttonTwo} onPress={ () => props.navigation.navigate('Home')} >
                    <Text style={Styles.buttonTextTwo}>CONTINUAR COMPRANDO</Text>
                </TouchableOpacity>

                <TouchableOpacity style={Styles.button} onPress={ () => setOpenModal(true)} >
                    <Text style={Styles.buttonText}>FINALIZAR COMPRA</Text>
                </TouchableOpacity>
            </View>
            
        </>

    )
}

export default ShoppingCart