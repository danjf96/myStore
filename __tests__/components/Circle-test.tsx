/**
 * @format
 */

import 'react-native';
import React from 'react';
import Circle from '../../src/components/Circle';
import { render } from 'react-native-testing-library';
import { Provider } from 'react-redux';
import store from '../../src/store';
import { CARTSTATE, Types } from '../../src/store/ducks/shoppingCart/types';

const component = (
    <Provider store={store} > <Circle /> </Provider>
);

const fakeValues:CARTSTATE[] = [{ 
    createdAt: "2019-09-02T12:58:54.103Z",
    id: "1",
    image: "http://lorempixel.com/640/480/food",
    name: "Rustic Metal Fish",
    price: 289.00,
    stock: 65171,
    number: 1
}]

describe('Render component', () => {
    it('renders correctly', async () => { 
        const screen = render(component)
        expect(screen.toJSON()).toMatchSnapshot()
    });

    test('render number props', () => {
        const screen = render(<Provider store={store} > <Circle number={2}/> </Provider>)
        expect(screen.getByText('2')).toBeTruthy()
    })

    test('render number', () => {
        const screen = render(component)
        store.dispatch({ type: Types.CHANGE, payload: { cart: fakeValues }})
        expect(screen.getByText('1')).toBeTruthy()
    })

})