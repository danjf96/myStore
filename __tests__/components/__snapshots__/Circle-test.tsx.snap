// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render component renders correctly 1`] = `
Array [
  " ",
  " ",
]
`;
